src/engine/shanghai.md

```diff
@@ -134,4 +148,61 @@ This method follows the same specification as [`engine_forkchoiceUpdatedV1`](./p
 
 This method follows the same specification as [`engine_getPayloadV1`](./paris.md#engine_getpayloadv1) with the addition of the following:
 
-  1. Client software **SHOULD** use the sum of the block's priority fees or any other algorithm to determine `blockValue`.
\ No newline at end of file
+  1. Client software **SHOULD** use the sum of the block's priority fees or any other algorithm to determine `blockValue`.
+
+### engine_getPayloadBodiesByHashV1
+
+#### Request
+
+* method: `engine_getPayloadBodiesByHashV1`
+* params:
+  1. `Array of DATA`, 32 Bytes - Array of `block_hash` field values of the `ExecutionPayload` structure
+* timeout: 10s
+
+#### Response
+
+* result: `Array of ExecutionPayloadBodyV1` - Array of [`ExecutionPayloadBodyV1`](#ExecutionPayloadBodyV1) objects.
+* error: code and message set in case an exception happens while processing the method call.
+
+#### Specification
+
+1. Given array of block hashes client software **MUST** respond with array of `ExecutionPayloadBodyV1` objects with the corresponding hashes respecting the order of block hashes in the input array.
+
+1. Client software **MUST** place responses in the order given in the request, using `null` for any missing blocks. For instance, if the request is `[A.block_hash, B.block_hash, C.block_hash]` and client software has data of payloads `A` and `C`, but doesn't have data of `B`, the response **MUST** be `[A.body, null, C.body]`.
+
+1. Client software **MAY NOT** respond to requests for finalized blocks by hash.
+
+1. Client software **MUST** set `withdrawals` field to `null` for bodies of pre-Shanghai blocks.
+
+1. This request maps to [`BeaconBlocksByRoot`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/p2p-interface.md#beaconblocksbyroot) in the consensus layer `p2p` specification. Callers must be careful to use the execution block hash, instead of the beacon block root.
+
+### engine_getPayloadBodiesByRangeV1
+
+#### Request
+
+* method: `engine_getPayloadBodiesByRangeV1`
+* params:
+  1. `start`: `QUANTITY`, 64 bits - Starting block number
+  1. `count`: `QUANITTY`, 64 bits - Number of blocks to return
+* timeout: 10s
+
+#### Response
+
+* result: `Array of ExecutionPayloadBodyV1` - Array of [`ExecutionPayloadBodyV1`](#ExecutionPayloadBodyV1) objects.
+* error: code and message set in case an exception happens while processing the method call.
+
+#### Specification
+
+1. Given a `start` and a `count`, the client software **MUST** respond with array of `ExecutionPayloadBodyV1` objects with the corresponding execution block number respecting the order of blocks in the canonical chain, as selected by the latest `engine_forkchoiceUpdated` call.
+
+1. Client software **MUST** support `count` values of at least 32 blocks.
```

Comment:
@arnetheduck while we're here, why 32 when `MAX_REQUEST_BLOCKS` is set to 1024?

Reply:
1024 was considered too big to stream in one response while 32 is the de-facto value that most clients use on the libp2p side for similar reasons (ie finality happens one epoch at a time, so most clients fetch one epoch at a time as well) - this is a .. pragmatic value more than anything else.

I do wonder if we need a mechanism for saying "request too large" akin to the above point about `null` and "I don't know" - ie if I request 1024 blocks and the response is 32 blocks and the EL knows there comes more after, should it signal this to the CL somehow? I'm open to suggestions here, will have a think about it otherwise - I think it's safe to clarify this point in a follow-up PR, it doesn't matter greatly from a practical pov as long as everyone just sticks to 32 as max request size